File: tool-ui/src/main/java/cn/lineai/tool/ui/DiffLines.java
package cn.lineai.tool.ui;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/** Bounded-memory line diff. Large replacement regions remain valid, non-minimal edits. */
public final class DiffLines {
    public static final class Line {
        public final int kind; // -1 removed, 0 unchanged, 1 added
        public final int number;
        public final String text;
        public final boolean terminated;
        Line(int kind, int number, String text) {
            this.kind = kind; this.number = number; this.terminated = text.endsWith("\n");
            this.text = terminated ? text.substring(0, text.length() - 1) : text;
        }
    }
    public final List<Line> lines;
    public final int added;
    public final int removed;
    private DiffLines(List<Line> lines) {
        this.lines = Collections.unmodifiableList(lines);
        int plus = 0, minus = 0;
        for (Line line : lines) { if (line.kind == 1) plus++; else if (line.kind == -1) minus++; }
        added = plus; removed = minus;
    }
    public static DiffLines calculate(String oldText, String newText) {
        String[] a = split(oldText), b = split(newText);
        int prefix = 0, suffix = 0;
        while (prefix < a.length && prefix < b.length && a[prefix].equals(b[prefix])) prefix++;
        while (suffix < a.length - prefix && suffix < b.length - prefix
                && a[a.length - suffix - 1].equals(b[b.length - suffix - 1])) suffix++;
        ArrayList<Line> lines = new ArrayList<>();
        for (int i = 0; i < prefix; i++) lines.add(new Line(0, i + 1, a[i]));
        int m = a.length - prefix - suffix, n = b.length - prefix - suffix;
        if ((long) (m + 1) * (n + 1) <= 1_000_000) {
            int[][] lcs = new int[m + 1][n + 1];
            for (int i = m - 1; i >= 0; i--) for (int j = n - 1; j >= 0; j--) {
                lcs[i][j] = a[prefix + i].equals(b[prefix + j]) ? 1 + lcs[i + 1][j + 1] : Math.max(lcs[i + 1][j], lcs[i][j + 1]);
            }
            int i = 0, j = 0;
            while (i < m || j < n) {
                if (i < m && j < n && a[prefix + i].equals(b[prefix + j])) {
                    lines.add(new Line(0, prefix + j + 1, b[prefix + j])); i++; j++;
                } else if (i < m && (j == n || lcs[i + 1][j] >= lcs[i][j + 1])) {
                    lines.add(new Line(-1, prefix + i + 1, a[prefix + i])); i++;
                } else { lines.add(new Line(1, prefix + j + 1, b[prefix + j])); j++; }
            }
        } else {
            for (int i = 0; i < m; i++) lines.add(new Line(-1, prefix + i + 1, a[prefix + i]));
            for (int j = 0; j < n; j++) lines.add(new Line(1, prefix + j + 1, b[prefix + j]));
        }
        for (int j = b.length - suffix; j < b.length; j++) lines.add(new Line(0, j + 1, b[j]));
        return new DiffLines(lines);
    }
    private static String[] split(String text) {
        if (text == null || text.isEmpty()) return new String[0];
        // A terminating newline is not an extra empty source line.
        String normalized = text.replace("\r\n", "\n");
        String[] lines = normalized.split("\n", -1);
        boolean terminated = normalized.endsWith("\n");
        if (terminated) lines = java.util.Arrays.copyOf(lines, lines.length - 1);
        for (int i = 0; i < lines.length; i++) if (i < lines.length - 1 || terminated) lines[i] += "\n";
        return lines;
    }
}

File: src/domain/diff_lines.cpp
#include "domain/diff_lines.h"

#include <algorithm>
#include <cstdint>
#include <string>
#include <vector>

namespace linecode::domain {
namespace {

// Mirrors the legacy `DiffLines.split`: CRLF is normalized, a terminating
// newline does not create a trailing empty line, and every line keeps its own
// terminator so `DiffLine::terminated` can be derived from it.
std::vector<std::string> SplitLines(const std::string_view text) {
  std::vector<std::string> lines;
  if (text.empty())
    return lines;
  std::string normalized{text};
  for (std::size_t index = 0; index + 1 < normalized.size();) {
    if (normalized[index] == '\r' && normalized[index + 1] == '\n')
      normalized.erase(index, 1);
    else
      ++index;
  }
  std::size_t start = 0;
  while (true) {
    const auto newline = normalized.find('\n', start);
    if (newline == std::string::npos) {
      if (start < normalized.size())
        lines.push_back(normalized.substr(start));
      break;
    }
    // Include the newline so the caller can detect the terminator.
    lines.push_back(normalized.substr(start, newline - start + 1));
    start = newline + 1;
    if (start >= normalized.size())
      break;
  }
  return lines;
}

DiffLine MakeLine(const DiffLine::Kind kind, const int number,
                  std::string text) {
  const bool terminated = text.ends_with('\n');
  if (terminated)
    text.pop_back();
  return DiffLine{.kind = kind,
                  .number = number,
                  .text = std::move(text),
                  .terminated = terminated};
}

} // namespace

DiffLines CalculateDiffLines(const std::string_view old_text,
                             const std::string_view new_text) {
  const auto a = SplitLines(old_text);
  const auto b = SplitLines(new_text);
  const auto a_size = static_cast<int>(a.size());
  const auto b_size = static_cast<int>(b.size());

  int prefix = 0;
  while (prefix < a_size && prefix < b_size && a[prefix] == b[prefix])
    ++prefix;
  int suffix = 0;
  while (suffix < a_size - prefix && suffix < b_size - prefix &&
         a[a_size - suffix - 1] == b[b_size - suffix - 1])
    ++suffix;

  DiffLines result;
  for (int index = 0; index < prefix; ++index)
    result.lines.push_back(
        MakeLine(DiffLine::Kind::unchanged, index + 1, a[index]));

  const int m = a_size - prefix - suffix;
  const int n = b_size - prefix - suffix;
  const auto cells = static_cast<std::int64_t>(m + 1) * (n + 1);
  if (cells <= 1'000'000) {
    // lcs[i][j] = LCS length of a[prefix+i..] and b[prefix+j..].
    std::vector<int> lcs(static_cast<std::size_t>(m + 1) *
                             static_cast<std::size_t>(n + 1),
                         0);
    const auto at = [n, &lcs](const int i, const int j) -> int & {
      return lcs[static_cast<std::size_t>(i) *
                     static_cast<std::size_t>(n + 1) +
                 static_cast<std::size_t>(j)];
    };
    for (int i = m - 1; i >= 0; --i) {
      for (int j = n - 1; j >= 0; --j) {
        at(i, j) = a[prefix + i] == b[prefix + j]
                       ? 1 + at(i + 1, j + 1)
                       : std::max(at(i + 1, j), at(i, j + 1));
      }
    }
    int i = 0;
    int j = 0;
    while (i < m || j < n) {
      if (i < m && j < n && a[prefix + i] == b[prefix + j]) {
        result.lines.push_back(
            MakeLine(DiffLine::Kind::unchanged, prefix + j + 1, b[prefix + j]));
        ++i;
        ++j;
      } else if (i < m && (j == n || at(i + 1, j) >= at(i, j + 1))) {
        result.lines.push_back(
            MakeLine(DiffLine::Kind::removed, prefix + i + 1, a[prefix + i]));
        ++i;
      } else {
        result.lines.push_back(
            MakeLine(DiffLine::Kind::added, prefix + j + 1, b[prefix + j]));
        ++j;
      }
    }
  } else {
    for (int i = 0; i < m; ++i)
      result.lines.push_back(
          MakeLine(DiffLine::Kind::removed, prefix + i + 1, a[prefix + i]));
    for (int j = 0; j < n; ++j)
      result.lines.push_back(
          MakeLine(DiffLine::Kind::added, prefix + j + 1, b[prefix + j]));
  }

  for (int j = b_size - suffix; j < b_size; ++j)
    result.lines.push_back(MakeLine(DiffLine::Kind::unchanged, j + 1, b[j]));

  for (const auto &line : result.lines) {
    if (line.kind == DiffLine::Kind::added)
      ++result.added;
    else if (line.kind == DiffLine::Kind::removed)
      ++result.removed;
  }
  return result;
}

} // namespace linecode::domain
